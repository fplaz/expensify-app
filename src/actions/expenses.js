// import { v4 as uuidv4 } from 'uuid'
import db from '../firebase/firebase'
import { ref, set, push, child, get, remove, update } from 'firebase/database';

// ADD EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        
        const expense = {description, note, amount, createdAt}
        
        const expenseRef = ref(db, 'users/'+ uid + '/expenses')
        const newExpenseRef = push(expenseRef)
        
        try {
            await set(newExpenseRef, expense)
            return dispatch(addExpense({
                id: newExpenseRef.key,
                ...expense
            }))
        } catch(e) {
            console.log(e)
        }
    }
}

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const StartEditExpense = (id, updates) => {
    return async (dispatc, getStateh) => {
        const uid = getState().auth.uid
        try {
            const dbRef = ref(db, 'users/' + uid + '/expenses/' + id)
            const expense = await update(dbRef, updates)
            return dispatch(editExpense(id, updates))
        }
        catch {
            console.log('StartEditExpense error:' + e)
        }
    }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid
        try {
            const expenses = []
            const dbRef = ref(db, 'users/' + uid + '/expenses')
            const snapshot = await get(dbRef)
            snapshot.forEach((item) => {
                expenses.push({
                    id: item.key,
                    ...item.val()
                })
            })
            return dispatch(setExpenses(expenses))
        } catch(e) {
            console.log('StartSetExpenses error:' + e)
        }
    }
}

// REMOVE EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = (id) => {
    return async (dispatch, getState) => {
        
        const uid = getState().auth.uid
        const expenseRef = ref(db, '/users/' + uid + '/expenses/' + id)

        try {
            await set(expenseRef, null)
            return await dispatch(removeExpense({id}))
        } catch(e) {
            console.log(e)
        }
    }
}