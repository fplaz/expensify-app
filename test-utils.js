import React from 'react'
import { render as rtlRender} from '@testing-library/react'
import configureStore from './src/store/configureStore' 
import { Provider } from 'react-redux'

function render(
    ui,
    {
      initialState = {},
      store = configureStore,
      ...renderOptions
    } = {}
  ) {
    const Wrapper = ({ children }) => {
      return <Provider store={store()}>{children}</Provider>;
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
  }
  
  // re-export everything
  export * from '@testing-library/react';
  
  // override render method
  export { render };