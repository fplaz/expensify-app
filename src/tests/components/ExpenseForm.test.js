import React from "react"
import { MemoryRouter } from 'react-router-dom'
import {render, screen, fireEvent} from '../../../test-utils'
import '@testing-library/jest-dom'
import ExpenseForm from "../../components/ExpenseForm"
import expenses from "../fixtures/expenses"

test('Should render expense form correctly', () => {
    const {asFragment} = render(
        <MemoryRouter>
            <ExpenseForm />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
})

test('Should render expense form with data correctly', () => {
    const {asFragment} = render(
        <MemoryRouter>
            <ExpenseForm expense={expenses[1]} />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
})

test('Should render error for invalid form submission', async () => {
    render(
        <MemoryRouter>
            <ExpenseForm />
        </MemoryRouter>
    )
    const submitButton = screen.getByText('Add Expense')
    fireEvent.click(submitButton)

    const errorMessage = screen.getByText('Please provide description and amount.')
    expect(errorMessage).toBeInTheDocument()
})

test('Should set description on input change', () => {
    render(
        <MemoryRouter>
            <ExpenseForm />
        </MemoryRouter>
    )

    const descriptionInput = screen.getByPlaceholderText("Description")
    fireEvent.change(descriptionInput, { target: { value: 'New Description' } })
    expect(screen.getByDisplayValue('New Description')).toBeInTheDocument()
})

test('Should call onSubmit prop for valid from submission', () => {
    const onSubmitSpy = jest.fn()
    render(
        <MemoryRouter>
            <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy}/>
        </MemoryRouter>
    )
    const submitButton = screen.getByText('Add Expense')
    fireEvent.click(submitButton)

    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[0].description,
        amount: expenses[0].amount,
        note: expenses[0].note,
        createdAt: expenses[0].createdAt.valueOf(),
    })
})