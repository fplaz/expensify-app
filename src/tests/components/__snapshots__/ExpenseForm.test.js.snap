// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense form correctly 1`] = `
<DocumentFragment>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        name="description"
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        name="amount"
        placeholder="Amount"
        type="text"
        value=""
      />
      <div
        class="SingleDatePicker SingleDatePicker_1"
      >
        <div>
          <div
            class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
          >
            <div
              class="DateInput DateInput_1"
            >
              <input
                aria-describedby="DateInput__screen-reader-message-date"
                aria-label="Date"
                autocomplete="off"
                class="DateInput_input DateInput_input_1"
                id="date"
                name="date"
                placeholder="Date"
                type="text"
                value="11/03/2023"
              />
              <p
                class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                id="DateInput__screen-reader-message-date"
              >
                Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
              </p>
            </div>
          </div>
        </div>
      </div>
      <textarea
        placeholder="Add a note for your expense (optional)"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Should render expense form with data correctly 1`] = `
<DocumentFragment>
  <div>
    <form
      data-testid="expense-form"
    >
      <input
        name="description"
        placeholder="Description"
        type="text"
        value="Rent"
      />
      <input
        name="amount"
        placeholder="Amount"
        type="text"
        value="1095"
      />
      <div
        class="SingleDatePicker SingleDatePicker_1"
      >
        <div>
          <div
            class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
          >
            <div
              class="DateInput DateInput_1"
            >
              <input
                aria-describedby="DateInput__screen-reader-message-date"
                aria-label="Date"
                autocomplete="off"
                class="DateInput_input DateInput_input_1"
                id="date"
                name="date"
                placeholder="Date"
                type="text"
                value="12/27/1969"
              />
              <p
                class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                id="DateInput__screen-reader-message-date"
              >
                Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
              </p>
            </div>
          </div>
        </div>
      </div>
      <textarea
        placeholder="Add a note for your expense (optional)"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;
